### PR TITLE
feat: initial implementation: page-level bundle size check for nextjs

### PR DESCRIPTION
**Pull request by masone**
calculates the size of a full page bundle based on the chunks in nextjs' `build-manifest.json`. generates a `bundlesize.config.json` for that and runs [bundlesize](https://github.com/siddharthkp/bundlesize) on it

RFC see https://github.com/carforyou/carforyou-docs/issues/96

```
npm install @carforyou/next-page-bundlesize -D
next build
npx next-page-bundlesize --maxSize="200 kB" --buildDir=.next
```

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,2 +1,83 @@
...
+interface Args {
+  maxSize: string
+  buildDir: string
+}
+
+interface Manifest {
+  pages: {
+    [pageName: string]: string[]
+  }
+  [key: string]: Record<string, unknown>
```
Comment (Averethel) [inline src/index.ts:15]: Any chance we can have a concrete type instead of `unknown`?

File: src/index.ts [hunk 2]
```diff
@@ -1,2 +1,83 @@
...
+    )
+      .filter((chunk) => chunk.match(/\.js$/))
+      .map((chunk) => path.join(buildDir, chunk))
+
+    const outFile = path.join(
+      buildDir,
+      `.bundlesize${page.replace(/[/]/g, "_").replace(/[[\]]/g, "-")}`
+    )
+
+    fs.writeFileSync(outFile, "")
```
Comment (Averethel) [inline src/index.ts:37]: Do we need to call `writeFile ` first? If I recall `appendFile` will create a file for you if it doesn't exist

File: src/index.ts [hunk 3]
```diff
@@ -1,2 +1,83 @@
...
+}
+
+export default function run() {
+  try {
+    const args = yargs.argv as unknown as Args
+    const maxSize = args.maxSize || "200 kB"
+    const buildDir = args.buildDir || ".next"
+
+    const manifestFile = path.join(buildDir, "build-manifest.json")
+    const manifest = JSON.parse(fs.readFileSync(manifestFile).toString())
```
Comment (Averethel) [inline src/index.ts:69]: Why do you need to convert to string here? Wouldn't 
```
JSON.parse(fs.readFileSync(manifestFile))
```
work?

File: src/index.ts [hunk 4]
```diff
@@ -1,2 +1,83 @@
...
+
+    const pageBundles = concatenatePageBundles({ buildDir, manifest })
+    const config = generateBundleSizeConfig({ pageBundles, maxSize })
+    const configFile = path.join(buildDir, "next-page-bundlesize.config.json")
+    fs.writeFileSync(configFile, config)
+
+    execSync(`npx bundlesize --config=${configFile}`, { stdio: "inherit" })
+    process.exit(0)
```
Comment (Averethel) [inline src/index.ts:77]: Don't we want to exit with code 1 if the bundlesize check fails? Exit codes are what CI looks for when checking for success if I recall correctly

**Review by Averethel — APPROVED** [review]
Nice! A few questions from my side, please take a look before merging

File: src/index.ts (hunk 2 above)
Comment (masone) [reply src/index.ts:37]: correct, but on subsequent builds, it might already exist. at least I ran into that case when testing manually, so I opted to be on the safe side

File: src/index.ts (hunk 3 above)
Comment (masone) [reply src/index.ts:69]: afaik, `JSON.parse` does not accept a stream as an argument https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse

File: src/index.ts (hunk 4 above)
Comment (masone) [reply src/index.ts:77]: `execSync` would throw and end up in the catch, where the process exits with 1. but to be a bit more on the defensive side, I can move it down to the end of the function

File: src/index.ts [hunk 5]
```diff
@@ -1,2 +1,84 @@
...
+}
+
+const concatenatePageBundles = ({
+  buildDir,
+  manifest,
+}: {
+  buildDir: string
+  manifest: Manifest
+}): string[] =>
+  Object.keys(manifest.pages).map((page) => {
```
Comment (lkappeler) [inline src/index.ts:25]: It's rather hard to follow, maybe you can split the concatenation of the chunks from the aggregation or add some comments.

**Review by lkappeler — APPROVED** [review]
Nice effort, maybe we wanna consider not adding the files to the build directory but in a separate folder, other than the more of a cosmetic input.

**Review by talamcol — APPROVED** [review]
Nice! what can we do if we have a page that is 231 kB (I saw the related listings PR with the 230 kB) and we want it to pass only for this single page? is there some manual exclude we can add?

**Comment by masone** [comment]
@talamcol there would be no such functionality. these should be considered hard limits. the problem on the page should be solved instead of raising the threshold. we can start high and continually aim for lowering it

File: src/index.ts (hunk 5 above)
Comment (masone) [reply src/index.ts:30]: never add comments when you can extract :D
